In [1]:
import pandas as pd 
import numpy as np

In [9]:
data=pd.read_csv('C:/Users/harsh/Desktop/insaid assignment/Fraud.csv',header=0)
data.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

In [12]:
data.describe()

step        amount  oldbalanceOrg  newbalanceOrig  \
count  6.362620e+06  6.362620e+06   6.362620e+06    6.362620e+06   
mean   2.433972e+02  1.798619e+05   8.338831e+05    8.551137e+05   
std    1.423320e+02  6.038582e+05   2.888243e+06    2.924049e+06   
min    1.000000e+00  0.000000e+00   0.000000e+00    0.000000e+00   
25%    1.560000e+02  1.338957e+04   0.000000e+00    0.000000e+00   
50%    2.390000e+02  7.487194e+04   1.420800e+04    0.000000e+00   
75%    3.350000e+02  2.087215e+05   1.073152e+05    1.442584e+05   
max    7.430000e+02  9.244552e+07   5.958504e+07    4.958504e+07   

       oldbalanceDest  newbalanceDest       isFraud  isFlaggedFraud  
count    6.362620e+06    6.362620e+06  6.362620e+06    6.362620e+06  
mean     1.100702e+06    1.224996e+06  1.290820e-03    2.514687e-06  
std      3.399180e+06    3.674129e+06  3.590480e-02    1.585775e-03  
min      0.000000e+00    0.000000e+00  0.000000e+00    0.000000e+00  
25%      0.000000e+00    0.000000e+00  0.000000e+00    0.000000e+00  
50%      1.327057e+05    2.146614e+05  0.000000e+00    0.000000e+00  
75%      9.430367e+05    1.111909e+06  0.000000e+00    0.000000e+00  
max      3.560159e+08    3.561793e+08  1.000000e+00    1.000000e+00

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


# Checking for null values

In [4]:
print(data.isnull().sum())

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64


# types of transactions

In [7]:
print(data.type.value_counts())

CASH_OUT    2237500
PAYMENT     2151495
CASH_IN     1399284
TRANSFER     532909
DEBIT         41432
Name: type, dtype: int64


# checking correlations between values

In [8]:
data.corr()

C:\Users\harsh\AppData\Local\Temp\ipykernel_42020\2627137660.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()


step    amount  oldbalanceOrg  newbalanceOrig  \
step            1.000000  0.022373      -0.010058       -0.010299   
amount          0.022373  1.000000      -0.002762       -0.007861   
oldbalanceOrg  -0.010058 -0.002762       1.000000        0.998803   
newbalanceOrig -0.010299 -0.007861       0.998803        1.000000   
oldbalanceDest  0.027665  0.294137       0.066243        0.067812   
newbalanceDest  0.025888  0.459304       0.042029        0.041837   
isFraud         0.031578  0.076688       0.010154       -0.008148   
isFlaggedFraud  0.003277  0.012295       0.003835        0.003776   

                oldbalanceDest  newbalanceDest   isFraud  isFlaggedFraud  
step                  0.027665        0.025888  0.031578        0.003277  
amount                0.294137        0.459304  0.076688        0.012295  
oldbalanceOrg         0.066243        0.042029  0.010154        0.003835  
newbalanceOrig        0.067812        0.041837 -0.008148        0.003776  
oldbalanceDest        1.000000        0.976569 -0.005885       -0.000513  
newbalanceDest        0.976569        1.000000  0.000535       -0.000529  
isFraud              -0.005885        0.000535  1.000000        0.044109  
isFlaggedFraud       -0.000513       -0.000529  0.044109        1.000000

In [15]:
data['type']=data['type'].map({'CASH_OUT':1,'PAYMENT':2,'CASH_IN':3,'TRANSFER':4,'DEBIT':5})
data.head()

step  type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1     2   9839.64  C1231006815       170136.0       160296.36   
1     1     2   1864.28  C1666544295        21249.0        19384.72   
2     1     4    181.00  C1305486145          181.0            0.00   
3     1     1    181.00   C840083671          181.0            0.00   
4     1     2  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

# deleting extra columns

In [17]:
del data['nameOrig']
del data['nameDest']

In [18]:
data.head()

step  type    amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0     1     2   9839.64       170136.0       160296.36             0.0   
1     1     2   1864.28        21249.0        19384.72             0.0   
2     1     4    181.00          181.0            0.00             0.0   
3     1     1    181.00          181.0            0.00         21182.0   
4     1     2  11668.14        41554.0        29885.86             0.0   

   newbalanceDest  isFraud  isFlaggedFraud  
0             0.0        0               0  
1             0.0        0               0  
2             0.0        1               0  
3             0.0        1               0  
4             0.0        0               0

# splitting into train AND test data

In [19]:
x= data.drop('isFraud',axis=1)
y=data['isFraud']
print(x,y)

         step  type      amount  oldbalanceOrg  newbalanceOrig  \
0           1     2     9839.64      170136.00       160296.36   
1           1     2     1864.28       21249.00        19384.72   
2           1     4      181.00         181.00            0.00   
3           1     1      181.00         181.00            0.00   
4           1     2    11668.14       41554.00        29885.86   
...       ...   ...         ...            ...             ...   
6362615   743     1   339682.13      339682.13            0.00   
6362616   743     4  6311409.28     6311409.28            0.00   
6362617   743     1  6311409.28     6311409.28            0.00   
6362618   743     4   850002.52      850002.52            0.00   
6362619   743     1   850002.52      850002.52            0.00   

         oldbalanceDest  newbalanceDest  isFlaggedFraud  
0                  0.00            0.00               0  
1                  0.00            0.00               0  
2                  0.00          

In [20]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test= train_test_split(x,y,test_size=0.2,random_state=1)

In [22]:
x_test.describe()

step          type        amount  oldbalanceOrg  \
count  1.272524e+06  1.272524e+06  1.272524e+06   1.272524e+06   
mean   2.432145e+02  2.055689e+00  1.797321e+05   8.349984e+05   
std    1.424401e+02  9.807926e-01  5.974620e+05   2.889064e+06   
min    1.000000e+00  1.000000e+00  0.000000e+00   0.000000e+00   
25%    1.550000e+02  1.000000e+00  1.337140e+04   0.000000e+00   
50%    2.390000e+02  2.000000e+00  7.505746e+04   1.405472e+04   
75%    3.350000e+02  3.000000e+00  2.087174e+05   1.074490e+05   
max    7.430000e+02  5.000000e+00  6.750076e+07   5.731626e+07   

       newbalanceOrig  oldbalanceDest  newbalanceDest  isFlaggedFraud  
count    1.272524e+06    1.272524e+06    1.272524e+06    1.272524e+06  
mean     8.563612e+05    1.099568e+06    1.223670e+06    7.858398e-07  
std      2.925002e+06    3.405392e+06    3.674981e+06    8.864761e-04  
min      0.000000e+00    0.000000e+00    0.000000e+00    0.000000e+00  
25%      0.000000e+00    0.000000e+00    0.000000e+00    0.000000e+00  
50%      0.000000e+00    1.322531e+05    2.142632e+05    0.000000e+00  
75%      1.448878e+05    9.435139e+05    1.112763e+06    0.000000e+00  
max      4.731626e+07    3.281945e+08    3.284317e+08    1.000000e+00

# training ML model

In [23]:
from sklearn.tree import DecisionTreeClassifier

In [26]:
model=DecisionTreeClassifier()
model.fit(x_train,y_train)
model.score(x_test,y_test)

0.9996958799991199

In [28]:
y_test_r=model.predict(x_test)
from sklearn.metrics import r2_score
r2_score(y_test,y_test_r)

0.7647228082775128

## model is able to explain 76.47% of variance and accuracy of 99.97%.